main.go

```diff
@@ -0,0 +1,706 @@
+	if !ok {
+		// You know what's cool in compiled languages? Type safety.
+		return false, errors.Errorf("expected an EdgeX Event, but got %T", event)
+	}
+
+	if len(event.Readings) < 1 {
+		// Is this really an error? EdgeX's Filter functions say yes.
+		return false, errors.New("event contains no Readings")
+	}
+
+	r := bytes.Buffer{}
+	decoder := json.NewDecoder(&r)
+	decoder.UseNumber()
+	decoder.DisallowUnknownFields()
+
+	for i := range event.Readings {
+		reading := &event.Readings[i] // Readings is 169 bytes. This avoid the copy.
+		switch reading.Name {
+		default:
+			app.lgr.Debug("Unknown reading.", "reading", reading.Name)
+			continue
+
+		case ResourceReaderNotification:
+			r.Reset()
+			r.WriteString(reading.Value)
+			notification := &llrp.ReaderEventNotification{}
+			if err := decoder.Decode(notification); err != nil {
+				app.lgr.Error("Failed to decode reader event notification", "error", err.Error())
+				continue
+			}
+
+			if err := app.handleReaderEvent(event.Device, notification); err != nil {
+				app.lgr.Error("Failed to handle ReaderEventNotification.",
+					"error", err.Error(), "device", event.Device)
+			}
+
+		case ResourceROAccessReport:
+			r.Reset()
+			r.WriteString(reading.Value)
+
+			report := &llrp.ROAccessReport{}
+			if err := decoder.Decode(report); err != nil {
+				app.lgr.Error("Failed to decode tag report",
+					"error", err.Error(), "device", event.Device)
+				continue
+			}
+
+			if report.TagReportData == nil {
+				app.lgr.Warn("No tag report data in report.", "device", event.Device)
+			} else {
+				app.reports <- reportData{report, inventory.NewReportInfo(reading)}
+				app.lgr.Trace("New ROAccessReport.",
+					"device", event.Device, "tags", len(report.TagReportData))
+			}
+		}
+	}
+
+	return false, nil
+}
+
+// handleReaderEvent handles an llrp.ReaderEventNotification from the Device Service.
+//
+// If a device reports a new connection event,
+// this adds the reader to the list of managed readers.
+// If a device reports a close event, it removes that reader.
+func (app *inventoryApp) handleReaderEvent(device string, notification *llrp.ReaderEventNotification) error {
+	const connSuccess = llrp.ConnectionAttemptEvent(llrp.ConnSuccess)
+
+	data := notification.ReaderEventNotificationData
+	switch {
+	case data.ConnectionAttemptEvent != nil && *data.ConnectionAttemptEvent == connSuccess:
+		return app.defaultGrp.AddReader(app.devService, device)
+
+	case data.ConnectionCloseEvent != nil:
+		app.defaultGrp.RemoveReader(device)
+	}
+
+	return nil
+}
+
+// writeInventorySnapshot writes the current inventory snapshot to w.
+func (app *inventoryApp) writeInventorySnapshot(w io.Writer) error {
+	// We send w and a writeErr channel into the inventory execution context
+	// and then wait to read a value from the writeErr channel.
+	// That context closes writeErr to signal the snapshot is written to w
+	// or an error prevented such, and we can send the result back to our caller.
```

Comment:
This comment tells me exactly what I can get from the code, but doesn't tell me why we are doing this.

Reply:
```go
// This is architected in a way that allows the calling routine to block until the request has
// been fulfilled by the main taskLoop in a thread-safe manner. This allows callers of the
// REST API to get a race-free result while also not impacting the performance of the
// processing logic (ie. thread preemption and mutex locking).
```

Reply:
Suggestion applied via #20